tools/testing/target_determination/heuristics/interface.py

```diff
@@ -0,0 +1,86 @@
+from abc import abstractmethod
+from typing import Any, Dict, List
+
+
+class TestPrioritizations:
+    """
+    Describes the results of whether heuristics consider a test relevant or not.
+    Heuristics can leave a list empty if they don't consider any tests relevant.
+
+    Heuristics can leave the unranked_relevance list empty to imply all unmentioned tests are irrelevant
+    """
+
+    highly_relevant: List[str]
+    probably_relevant: List[str]
+    unranked_relevance: List[
+        str
+    ]  # When we don't know if the test is relevant to the PR or not
+    # future cateories could include 'definitely not relevant' and 'probably not relevant'
+
+    def __init__(self) -> None:
+        self.highly_relevant = []
+        self.probably_relevant = []
+        self.unranked_relevance = []
+
+    @staticmethod
+    def _merge_tests(
+        current_tests: List[str], new_tests: List[str], exclude_tests: List[str]
+    ) -> List[str]:
+        """
+        We append all new tests to the current tests, while preserving the sorting on the new_tests
+        However, exclude any specified tests which have now moved to a higher priority list
+        """
+        current_tests.extend(new_tests)
+        current_tests = [
+            test for test in current_tests if test not in exclude_tests
+        ]  # skip the excluded tests
+        current_tests = list(dict.fromkeys(current_tests))  # remove dupes
+
+        return current_tests
+
+    def integrate_priorities(self, other: "TestPrioritizations") -> None:
+        """
+        Integrates priorities from another TestPrioritizations object.
+
+        Assumes tests are only shuffled around beteen the lists, with no tests added or removed.
+        """
+        # only add new tests to the list, while preserving the sorting
+        self.highly_relevant.extend(
+            [test for test in other.highly_relevant if test not in self.highly_relevant]
+        )
+
+        self.highly_relevant = TestPrioritizations._merge_tests(
+            self.highly_relevant, other.highly_relevant, exclude_tests=[]
+        )
+
+        self.probably_relevant = TestPrioritizations._merge_tests(
+            self.probably_relevant,
+            other.probably_relevant,
+            exclude_tests=self.highly_relevant,
+        )
+
+        # We don't expect every heuristics to list all unranked tests. Easier to compute the list from scratch
+        ranked_tests = self.highly_relevant + self.probably_relevant
+        self.unranked_relevance = [
+            t for t in self.unranked_relevance if t not in ranked_tests
+        ]
```

Comment:
I think it might be worth adding an assertion somewhere that total tests at beginning = total tests at the end